File: backend/app/routers/reports.py
from datetime import date

from fastapi import APIRouter, Depends, Query

from app.database import get_db
from app.dependencies.auth import get_current_admin
from app.schemas.report import (
    LowStockResponse,
    MovementHistoryResponse,
    StockSummaryResponse,
    TopProductsResponse,
)
from app.services.report_service import ReportService

router = APIRouter(prefix="/reports", tags=["Reports"])


@router.get("/stock-summary", response_model=StockSummaryResponse)
async def stock_summary(
    db=Depends(get_db),
    current_admin=Depends(get_current_admin),
):
    service = ReportService(db)
    return await service.get_stock_summary()


@router.get("/low-stock", response_model=LowStockResponse)
async def low_stock(
    db=Depends(get_db),
    current_admin=Depends(get_current_admin),
):
    service = ReportService(db)
    return await service.get_low_stock()


@router.get("/top-products", response_model=TopProductsResponse)
async def top_products(
    db=Depends(get_db),
    current_admin=Depends(get_current_admin),
):
    service = ReportService(db)
    return await service.get_top_products()


@router.get("/movement-history", response_model=MovementHistoryResponse)
async def movement_history(
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=100),
    start_date: date | None = Query(None),
    end_date: date | None = Query(None),
    db=Depends(get_db),
    current_admin=Depends(get_current_admin),
):
    service = ReportService(db)
    return await service.get_movement_history(
        page=page,
        page_size=page_size,
        start_date=start_date,
        end_date=end_date,
    )

File: frontend/src/api/reports.ts
import client from './client'
import type { MovementType } from './stock'

export type StockSummaryItem = {
  id: string
  name: string
  sku: string
  quantity: number
  price: string
  inventory_value: string
  is_low_stock: boolean
  low_stock_threshold: number
  is_active: boolean
}

export type StockSummaryResponse = {
  items: StockSummaryItem[]
  total_inventory_value: string
  total_products: number
}

export type LowStockItem = {
  id: string
  name: string
  sku: string
  quantity: number
  low_stock_threshold: number
  is_low_stock: boolean
  is_active: boolean
}

export type LowStockResponse = {
  items: LowStockItem[]
  total: number
}

export type TopProductItem = {
  product_sku: string
  product_name: string
  total_quantity: number
  total_orders: number
  total_revenue: string
}

export type TopProductsResponse = {
  items: TopProductItem[]
  total: number
}

export type MovementHistoryItem = {
  id: string
  product_id: string | null
  product_sku: string
  movement_type: MovementType
  quantity_change: number
  quantity_before: number
  quantity_after: number
  note: string | null
  created_by: string | null
  created_at: string
}

export type MovementHistoryResponse = {
  items: MovementHistoryItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export const getStockSummary = async (): Promise<StockSummaryResponse> => {
  const response = await client.get('/reports/stock-summary')
  return response.data as StockSummaryResponse
}

export const getLowStock = async (): Promise<LowStockResponse> => {
  const response = await client.get('/reports/low-stock')
  return response.data as LowStockResponse
}

export const getTopProducts = async (): Promise<TopProductsResponse> => {
  const response = await client.get('/reports/top-products')
  return response.data as TopProductsResponse
}

export const getMovementHistory = async (
  page = 1,
  pageSize = 10,
  startDate?: string,
  endDate?: string,
): Promise<MovementHistoryResponse> => {
  const response = await client.get('/reports/movement-history', {
    params: {
      page,
      page_size: pageSize,
      ...(startDate && { start_date: startDate }),
      ...(endDate && { end_date: endDate }),
    },
  })
  return response.data as MovementHistoryResponse
}
